### PR TITLE
story(gen-go): cover every discriminator path the automaton carries, as a whole file

### LINKED ISSUE #265: story(gen-go): cover every discriminator path the automaton carries, as a whole file
### Description

The file tier. #264 shows an adopter one record's bytes; this shows them a
**file** — framing and all, records in the order the automaton admits them — and
covers every discriminator path the automaton carries.

For each path, one test holding the whole file inline: read it with the generated
`Reader`, assert the record types come back in the expected order with the
expected values, write them back with the generated `Writer`, assert the bytes
are the bytes it started from.

This is where the adopter's real question gets answered. A record literal shows
whether a field is at the offset they think. A *file* literal shows whether the
records are the records they have, in the order they come in, told apart on the
bytes they are actually told apart on, behind the framing their dataset uses.

### Coverage is every discriminator path, not one happy path

Decided in #263: every transition predicate and every literal of a
`BytesOneOf` is satisfied by some case. A layout with six posting types produces
cases that between them admit all six; a predicate written as
`is one of "DA" or "DB"` produces a case for each literal, because those are two
ways a file can be spelled and an adopter checking one has not checked the other.

The set to solve over is small and closed, which is what makes this a bounded
search rather than a solver. Predicates are `BytesEqual` and `BytesOneOf`
(`ir.proto:747`, `:767`). Guards are `equals`, `one_of` and `greater_than_zero`
(`ir.proto:911`). Bindings are from a field of the admitted record, or
`Decrement` (`ir.proto:875`). There is no fourth of anything.

**Transitions that share a predicate are covered once, by predicate, not once per
edge.** This needs deciding rather than assuming, because it is the difference
between a readable file and an unreadable one:
[`example/graph/graph.md`](../example/graph/graph.md) draws nine states and fifty
transitions for one ordinary ledger, and every posting state offers all six
posting types, so a full edge cover is fifty cases for the worked example alone.
What an adopter is checking is *this record type is selected on these bytes* —
which is a property of the predicate — and the guards that make the edges
distinct are the automaton's business, drawn already by `cpybkc-gen-graph`.

### The registers are the hard part

A transition may bind an integer register from a count field and guard a later
one on it. `example/ledger.sexpr` is exactly that shape: the header states how
many postings follow, `times` binds `HDR-COUNT` into a register, each posting
edge is guarded `greater than zero` and decrements it, and the trailer's edge is
guarded on zero.

Synthesizing a file for that means the count field and the number of records
emitted after it have to agree — and the synthesizer has to **choose the count
first and store it**, because `docs/ir/SPEC.md`'s *"A writer evaluates a guard,
it never back-fills a count"* means nothing downstream will fix it up. Getting it
wrong does not produce a subtly wrong file; it produces a file the generated
reader rejects, which the golden packages will catch on the first run.

`example/README.md` already names both ends of what this has to get right:

> a file whose header says six and whose body holds five is reported rather than
> returned as the five records it managed, and a writer closed a posting short of
> its own header count is reported rather than emitting the file its reader would
> have complained about one build later.

### Framing is on the path, and there are four of them

Every case carries real framing, because the framing is half of what an adopter
is checking — a record descriptor word is four bytes they will see in a hexdump
and want to recognise. `docs/ir/SPEC.md`'s *"Four framings, and none of them is a
RECFM"* and *"Terminator, separator, and the last record"* are what has to be
emitted, and the six golden packages are already exactly that coverage: `orders`
descriptor-word, `fixed` unframed, `chunks` segmented, `counted` delimited with a
terminator, `sep` delimited with a separator, `opt` delimited with an optional
terminator.

That means this story needs no new fixture. If the file tier works, six packages
regenerate and their generated tests pass; if it does not, the framing that broke
is named by the package that failed.

### What is not settled

- **How the paths are chosen.** Shortest-first until every predicate is covered
  is the obvious rule and produces short, readable files. Whether a path that
  covers three new predicates beats three paths covering one each is a legibility
  call, not a correctness one, and is better made against the regenerated
  goldens.
- **How many repetitions a counted run gets.** One is enough to cover the
  predicate and shortest to read; two is what shows a reader that the loop is a
  loop and that the count is decremented rather than tested once. Two is the
  recommendation, and it is cheap to change.
- **What a case is named.** The name is what a reader sees first in a failure and
  is the only summary of the path it walks. `TestAHeaderThenTwoDebitPostingsThenATrailer`
  is legible and long; a numbered name is short and says nothing.
- **Whether the file tier asserts field values too, or only record types and
  order.** #263 settled *round-trip plus field values* for a case; at the file
  tier that could mean every field of every record, which for the ledger's six
  posting types is a large file. Asserting the discriminated field plus byte
  equality, and leaving the full field assertions to the record tier where each
  record already has a case of its own, is the recommendation — but it is a
  narrowing of #263's decision and has to be taken deliberately.
- **A descriptor whose automaton admits no record** produces no `file.go` today,
  per `cmd/cpybkc-gen-go/README.md`. It should produce no file-tier test either,
  for the same reason, and that should be said rather than left to fall out.

### Acceptance Criteria

- [ ] `cpybkc-gen-go` writes a file-tier test file into `--out`, under the name
      #263 fixed, alongside the record tier #264 emits
- [ ] The cases between them satisfy **every** transition predicate in the
      descriptor and **every** literal of a `BytesOneOf`
- [ ] A test in `cmd/cpybkc-gen-go` holds the generator to that coverage rule, so
      a descriptor whose predicates are not all reached fails rather than
      generating a quiet gap
- [ ] Transitions sharing a predicate are covered once, by predicate; the
      decision and its reason are written into `cmd/cpybkc-gen-go/README.md`
      because it is a bound on what the tests prove
- [ ] Registers are satisfied: a count field bound into an integer register
      carries the number of records the path actually emits, chosen before it is
      stored, and a guarded exit is reached with the register in the state its
      guard requires
- [ ] Each case carries the framing inline and visible — a record descriptor
      word, a delimiter, a segment header — not stripped or elided
- [ ] All four framings and all three delimiter placements are exercised by the
      six golden packages regenerating and their generated tests passing
- [ ] Each case reads with the generated `Reader`, asserts the record types come
      back in the expected order, writes them back with the generated `Writer`
      and asserts byte equality
- [ ] The style rules hold here too: bytes inline, no helper shared between
      cases, no package-level state, one top-level test function per path
- [ ] A descriptor whose automaton admits no record produces no file-tier test,
      the way it already produces no `file.go`
- [ ] Output is deterministic: the path chosen for a descriptor does not depend
      on map iteration order, the clock, or anything outside the descriptor and
      the options
- [ ] The six goldens and `example/ledger/` are regenerated and checked in, and
      the generated file-tier tests pass under `go test -race ./...`
- [ ] `example/graph/graph.md` and the regenerated `example/ledger` file-tier test
      describe the same automaton — the diagram and the bytes are the two halves
      of the same spot-check and a reviewer should be able to hold one against
      the other

### Related Issues

- #263 decides the file name, package clause, spelling and style; #264 lands
  the synthesizer and the spelling this reuses for the records inside a file
- #52 emits the file-level reader and writer these cases drive, and is where the
  framing and the automaton walk live
- #189 is `cpybkc-gen-graph` drawing the same automaton; the diagram answers
  *which records, in which order, told apart on which bytes* about the descriptor
  and this answers it about the bytes
- #151 is where a predicate reading past the shortest record in front of it was
  settled — the guarantee that makes a synthesized file safe to build predicate
  by predicate
- #266 owns what happens when no accepting path exists at all

### PR DESCRIPTION
**Pull request by Zaba505**
Closes #265

The file tier of the generated tests. `cpybkc-gen-go` now writes `file_test.go`
beside `records_test.go`: one case per path through the automaton, each holding
a whole file inline — the framing as well as the records — read with the
generated `Reader`, checked for record types and order, written back with the
generated `Writer` and checked byte for byte.

## Acceptance criteria

- [x] `cpybkc-gen-go` writes a file-tier test file into `--out`, under the name #263 fixed (`file_test.go`), alongside the record tier #264 emits
- [x] The cases between them satisfy **every** transition predicate in the descriptor and **every** literal of a `BytesOneOf`
- [x] A test in `cmd/cpybkc-gen-go` holds the generator to that coverage rule — `TestTheFileTierCoversEveryTransitionPredicate` over all six golden descriptors, `TestEveryLiteralOfASetMembershipPredicateGetsACase` for the set half, and `TestAPredicateNoFileReachesIsRefused` for the gap
- [x] Transitions sharing a predicate are covered once, by predicate; the decision and its reason are in `cmd/cpybkc-gen-go/README.md`
- [x] Registers are satisfied: a count field bound into an integer register carries the number of records the path actually emits, chosen before it is stored, and a guarded exit is reached with the register in the state its guard requires
- [x] Each case carries the framing inline and visible — record descriptor word, delimiter, segment header — never stripped
- [x] All four framings and all three delimiter placements are exercised by the six golden packages regenerating and their generated tests passing
- [x] Each case reads with the generated `Reader`, asserts the record types come back in the expected order, writes them back with the generated `Writer` and asserts byte equality
- [x] The style rules hold: bytes inline, no helper shared between cases, no package-level state, one top-level test function per path
- [x] A descriptor whose automaton admits no record produces no file-tier test — `TestADescriptorWhoseAutomatonAdmitsNoRecordWritesNoFileTier`
- [x] Output is deterministic — `TestTheFileTierIsTheSameBytesTwice`
- [x] The six goldens and `example/ledger/` are regenerated and checked in, and the generated file-tier tests pass under `go test -race ./...`
- [x] `example/graph/graph.md` and the regenerated `example/ledger` file tier describe the same automaton, and `example/README.md` now says how to hold one against the other

## The judgment calls the story left open

**How the paths are chosen.** Shortest first: the shortest walk from the start
state to a state offering the transition being covered, the transition, and the
shortest walk from where it lands to a state that accepts. Every predicate a
path passes through is struck off with it, so the ledger gets **seven** cases
rather than fifty.

**How many repetitions a counted run gets.** Two, as recommended. It is what
tells a register that was decremented from one that was tested, and it is what
puts a separator in a delimited file at all — a file of one record under a
separator placement carries none, which is half the framing the goldens exist to
exercise. One constant, one line to change.

**What a case is named.** After the file it holds, with runs collapsed:
`TestALedgerHeaderThenTwoDebitPostingsThenALedgerTrailer`.

**Whether the file tier asserts field values.** Record types, their order, the
field each transition's predicate names, and byte equality over the whole file —
the narrowing the story recommended, taken deliberately and written up in
`README.md`. Every field of every record is already asserted at the record tier,
one file over.

**Coverage is by predicate, not by edge.** The reason is in the README: what an
adopter checks is *this record type is selected on these bytes*, which is a
property of the predicate; the guards that distinguish two edges carrying one
predicate are the automaton's business, and `cpybkc-gen-graph` has drawn them.

## The registers

Every guard the path evaluates and every table whose count is a register becomes
one requirement on the value a binding put in a register, shifted by the
decrements standing between the two. The set is closed — two predicate forms,
three guard tests, two binding forms — so each binding is settled on its own: the
requirements naming a set intersect, the requirements naming a bound narrow, and
the smallest value satisfying both wins. A binding nothing constrains is left
alone and its field takes the record tier's own derived value. A field the
admitting predicate already pins is *read* rather than chosen.

Every candidate path is then checked against the walk the generated reader would
actually make: both directions take the first eligible transition of a state, so
a path is only a path if every earlier transition is excluded — by a guard that
does not hold, or by a predicate the record's own bytes do not satisfy.

## Verified

`dagger call ci` green from an ordinary clone of this branch (a worktree's `.git`
is a file, so dagger cannot run there), plus `go test -race ./...`,
`golangci-lint run` and `gofmt -l` clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: cmd/cpybkc-gen-go/filetest.go [hunk 1]
```diff
@@ -0,0 +1,1855 @@
...
+
+	for at, taken := range walk {
+		for _, id := range t.walks[taken.from][taken.at].node.GetBindingIds() {
+			binding := t.nodes[id].GetBinding()
+			if _, takes := binding.GetValue().(*irpb.Binding_Decrement); !takes {
+				continue
+			}
+
+			for _, one := range sites {
+				if one.register == binding.GetRegisterId() && one.step <= at {
```
[the remark this answers is not in the document]
Comment (Zaba505) [reply cmd/cpybkc-gen-go/filetest.go]: Fixed, and thank you — this was real. `reachedWith` is now one forward walk in step order, the shape `demands` already used: a binding replaces what the register holds and resets the decrements with it, and a decrement takes one off whatever the latest binding put there.

Since no golden descriptor walks a path that rebinds a register, the goldens would have agreed with the bug. `TestADecrementCountsAgainstTheLatestBindingOfItsRegister` pins it directly on a hand-built header → detail → header → detail path over the `counted` automaton: the counter reads 0 at the end, where counting both decrements against the first header gave −1.

File: cmd/cpybkc-gen-go/filetest.go [hunk 2]
```diff
@@ -0,0 +1,1855 @@
...
+// segmented is one record cut into as few segments as this file node's largest
+// segment allows, each behind the segment descriptor word describing it.
+//
+// The only framing that stands *inside* a record rather than around it, so the
+// record's runs are cut at the segment boundaries and a run the cut falls inside
+// is split. The half behind the boundary keeps the item's note and says it is a
+// continuation, because a reader counting bytes against a hexdump needs both
+// halves labelled.
+func (t *filetest) segmented(at int, parts []chunk, width int) []chunk {
+	most := int(t.maxSegment) - segmentDescriptorWidth
```
[the remark this answers is not in the document]
Comment (Zaba505) [reply cmd/cpybkc-gen-go/filetest.go:1588]: This one is already refused upstream, before anything is laid out. `fileTests` calls `filer.collect`, which calls `filer.framing`, and that refuses a segmented file node whose largest segment is four bytes or fewer:

```
if f.maxSegment <= segmentDescriptorWidth {
    return malformed(... "a segment carries a four-byte segment descriptor word and the data behind it, so a largest segment of four or fewer carries no data at all")
}
```

So `most` is at least one by the time `segmented` runs. You could not see that from the diff — `file.go` is unchanged by this pull request — and the reasoning was right. I have left it as the one refusal rather than adding a second unreachable branch, on this repository’\s rule against code no descriptor can reach.

File: cmd/cpybkc-gen-go/filetest.go [hunk 3]
```diff
@@ -0,0 +1,1855 @@
...
+// four bytes they will see in a hexdump and want to recognise.
+func (t *filetest) framed(at int, parts []chunk) []chunk {
+	width := 0
+	for _, part := range parts {
+		width += len(part.body)
+	}
+
+	switch t.how {
+	case descriptorWord:
+		stated := width + segmentDescriptorWidth
```
[the remark this answers is not in the document]
Comment (Zaba505) [reply cmd/cpybkc-gen-go/filetest.go:1531]: Half taken.

The overflow check is in: a stated length that will not fit two bytes is now `malformed`, naming the record and its width, rather than a silently truncated word that fails byte equality inside generated code somebody else has to debug. That was the right call.

The constant is deliberate rather than wrong. `segmentDescriptorWidth` is documented in `file.go` as "the width of the segment descriptor word DFSMS defines, and of the record descriptor word beside it. Both are four bytes and the IR carries neither" — and the emitted writer sizes its own record descriptor word from that same constant. Two constants of one value would be the way the two arithmetics come apart, which is exactly what the doc comment above `framed` is guarding against.

File: cmd/cpybkc-gen-go/filetest.go [hunk 4]
```diff
@@ -0,0 +1,1855 @@
...
+		funcs []string
+		used  = make(map[string]struct{})
+	)
+
+	for _, one := range files {
+		funcs = append(funcs, t.testCase(one, alias, used))
+	}
+
+	paths := []string{"bytes", "errors", "io", "testing", t.opts.importPath}
+	for path := range t.synth.needs {
```
[the remark this answers is not in the document]
Comment (Zaba505) [reply cmd/cpybkc-gen-go/filetest.go]: Fixed both halves. The imports are collected per **accepted** case now — `laid` carries its own `needs`, and `write` takes a fresh `needs` map off the shared synthesizer around each candidate — so an import a discarded candidate reached for cannot reach the file. And the seed plus the per-case needs go through a set before they are sorted, so a duplicate cannot either.

Worth saying that today `needs` only ever holds `math/big`, and only from a `*big.Int` field assertion — but the file tier asserts discriminated fields, a discriminated field can be a numeric item wide enough for one, and "only reachable one way today" is not a reason to leave output that would not compile.

File: cmd/cpybkc-gen-go/filetest.go [hunk 5]
```diff
@@ -0,0 +1,1855 @@
...
+
+// paths is every candidate file for one goal, in the order they are preferred.
+//
+// A candidate is three pieces: the shortest walk from the start state to a state
+// offering the transition being covered, that transition, and the shortest walk
+// from where it lands to a state that accepts. Between the second and the third
+// the same transition is taken again where the state it lands in offers one
+// carrying the same predicate — see [run] for why twice rather than once, and
+// why the singular is tried after it rather than instead of it.
+func (t *filetest) paths(want goal) ([][]step, error) {
```
[the remark this answers is not in the document]
Comment (Zaba505) [reply cmd/cpybkc-gen-go/filetest.go:488]: Taken, with one qualification.

Candidates are now sorted by the length of the file they produce, so a goal two states offer is answered by the shorter file rather than by whichever state the descriptor carries first. The qualification is that they are sorted in two lists rather than one: the run walked `runLength` times is what this generator means to write, and the same run walked once is the *fallback* for when the longer file cannot be laid out. Sorting them together made a one-record `sep` file win over the two-record one — which loses the separator entirely, since a separator only exists between two records.

So: preferred candidates ordered by length, then shortened ones ordered by length. README.md now says so as well.

File: cmd/cpybkc-gen-go/filetest.go [hunk 6]
```diff
@@ -0,0 +1,1855 @@
...
+
+// eligible is whether one transition would take the record in front of it.
+func (t *filetest) eligible(edge transition, walk []step, sites []site, raw []byte) (bool, error) {
+	reached := make(map[uint64]reachedRegister)
+	for _, one := range t.reachedWith(walk, sites) {
+		reached[one.register] = one
+	}
+
+	for _, id := range edge.node.GetGuardIds() {
+		guard := t.nodes[id].GetGuard()
```
[the remark this answers is not in the document]
Comment (Zaba505) [reply cmd/cpybkc-gen-go/filetest.go]: Fixed. `eligible` resolves the guard node and returns `unresolved(id)` (or `malformed` for a node that is not a guard), and `reachedWith` does the same for a binding — matching what `needOf` and `bind` already did. A dangling reference is now the diagnostic it is rather than a coverage failure reported somewhere else entirely.

File: cmd/cpybkc-gen-go/filetest.go [hunk 7]
```diff
@@ -0,0 +1,1855 @@
...
+			lower = max(lower, shift(want.atLeast, want.down))
+			upper = min(upper, shift(want.atMost, want.down))
+			table = table || want.table
+
+			continue
+		}
+
+		shifted := make([]int64, 0, len(want.numbers))
+		for _, value := range want.numbers {
+			shifted = append(shifted, value+int64(want.down))
```
[the remark this answers is not in the document]
Comment (Zaba505) [reply cmd/cpybkc-gen-go/filetest.go]: Fixed — it goes through `shift` now, like the bounded path.

File: cmd/cpybkc-gen-go/filetest.go [hunk 8]
```diff
@@ -0,0 +1,1855 @@
...
+			continue
+		}
+
+		node, ok := t.nodes[met.predicate]
+		if !ok {
+			continue
+		}
+
+		values, err := t.literalsOf(met.predicate)
+		if err != nil || met.pick >= len(values) {
```
[the remark this answers is not in the document]
Comment (Zaba505) [reply cmd/cpybkc-gen-go/filetest.go]: Taken. `doc` returns an error, and so does `testCase` above it. An unresolved predicate is `unresolved(id)`, and a `pick` outside the set is `malformed` naming the literal index and how many the predicate carries.

Your reasoning for it is the right one: both invariants held once already in `goals()`, so either failing here means `met` and `goals` disagree about which literal a case covers — which is the gap the coverage rule exists to catch, and a comment that quietly grew shorter is the one outcome that would hide it.

File: cmd/cpybkc-gen-go/filetest_test.go [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+
+			reached := make(map[goal]struct{})
+
+			for _, file := range files {
+				for _, met := range file.meets {
+					reached[met] = struct{}{}
+				}
+			}
+
+			for _, missing := range want {
```
[the remark this answers is not in the document]
Comment (Zaba505) [reply cmd/cpybkc-gen-go/filetest_test.go:89]: Taken, and rewritten to the shape you point at.

`TestTheFileTierCoversEveryTransitionPredicate` now asserts over the **bytes**: for every goal, some accepted case must spell that literal at the offset that predicate reads it from, in a record the transition admits. It also calls `fileTests` and counts `func Test` against the number of paths chosen, which is the "chosen but never written" gap — the `enc == nil` return, the import assembly, `testCase` — that the old shape could not see.

You were right that the old one was close to tautological: `cover` already refuses a goal it cannot reach, so asking it which goals it reached asked it to mark its own work.

File: cmd/cpybkc-gen-go/filetest_test.go [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+				}
+			}
+		})
+	}
+}
+
+// newFileTest is the emitter the two coverage assertions reach past
+// [fileTests] for, so that they can ask what the goals were as well as what was
+// written.
+func newFileTest(d *irpb.Descriptor, opts options) (*filetest, error) {
```
[the remark this answers is not in the document]
Comment (Zaba505) [reply cmd/cpybkc-gen-go/filetest_test.go]: Fixed the way you suggest. There is now one `newFiletest` constructor: it indexes the descriptor, resolves every state's transitions and builds the synthesizer, and `fileTests` is that plus the emission rules and nothing else — no automaton, no encoding, no `--opt import_path`. The tests call the same constructor the generator does, so there is no second copy to drift.

The shadowing is gone too: `filetest.collect` is now `demands` (named for what it collects), and `source`'s local `paths` is `sorted`.

**Comment by Zaba505** [comment]
Thanks — six of the ten landed as changes, and the decrement one was a real bug that no golden could have caught. Answers to the two general points that were not inline:

**The delimiter question.** The generated reader never searches for the delimiter, so a record whose own bytes contain it is fine. `file.go` says so where it declares the constant: *"It is never searched for, either — a record's end comes from its extent, and 0x15 sits inside any COMP-3 field holding a value like +152.50."* The `counted` golden is built for exactly that trap — it is delimited by `0x15` and its detail record carries a `PIC S9(3)V99 COMP-3` field — and its generated file tier reads and writes back a `0x00 0x00 0x2d` payload behind a `0x15` terminator without trouble. Good instinct; the answer is that the framing layer is extent-driven on both sides.

**`pin()` clamping an out-of-range `pick`.** Left as a clamp, unlike `doc()`. `doc` is reporting what a case covers, so a disagreement there is a coverage claim that is wrong; `pin` is choosing bytes for a field the predicate already fixes, and the same clamp lives in `synth.require` for the record tier, where index 0 is the documented answer. Making one of the two loud and leaving the other silent would be worse than either, and moving both is a change to the record tier's settled behaviour that does not belong in this pull request.

Everything else is in the two commits above: goldens regenerate byte-identically after the refactor, `go test -race ./...`, `golangci-lint run` and `dagger call ci` all green.
